src/gadgets/nonnative/bn/big_uint_mul_mod_chip/mod.rs

```diff
@@ -941,6 +941,148 @@ pub struct SumContext<F: PrimeField> {
 }
 
 impl<F: ff::PrimeField> BigUintMulModChip<F> {
+    /// Converts a single assigned cell to a list of limbs, according to the limb width and count limit.
+    ///
+    /// This function assigns to the region context, evaluated `AssignedCells` representing the limbs of the
+    /// original `AssignedCell` in the `BigUint` representation.
+    ///
+    /// # Arguments
+    /// * `ctx`: Mutable reference to the `RegionCtx` provides the constraint system and metadata.
+    /// * `input`: A reference to the `AssignedCell` representing the input value.
+    ///
+    /// For every `k` rows looks like:
+    /// ```markdown
+    /// |   ---    |  ---  |  ---  |  ---  |  ---  |  ---   |
+    /// | state[0] | q1[0] |  q_i  | input |  q_o  | output |
+    /// |   ---    |  ---  |  ---  |  ---  |  ---  |  ---   |
+    /// |   ...    |  ...  |  ...  |  ...  |  ...  |  ...   |
+    /// |  limb_k  |   1   | shift |  p_k  |  -1   |  s_k   |
+    /// |   ...    |  ...  |  ...  |  ...  |  ...  |  ...   |
+    /// ```
+    /// where:
+    ///     - `limb_k` it's limb of input in big uint representation
+    ///     - `shift` it's constant equal `2 ^ self.limb_width`
+    ///     - `p_k` partial sum of [0..k-1] limbs
+    ///     - `s_k` partial sum of [0..k] limbs
+    ///
+    /// The `s_n` it's equal with original input and check by copy constraint
+    pub fn from_assigned_cell_to_limbs(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        input: &AssignedCell<F, F>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        trace!("Input: {:?}", input.value().unwrap());
+        let bignat_limb_column = &self.config().state[0];
+        let bignat_limb_shift = &self.config().q_1[0];
+
+        let prev_column = &self.config().input;
+        let prev_selector = &self.config().q_i;
+
+        let partial_sum_column = &self.config().out;
+        let partial_sum_selector = &self.config().q_o;
+
+        let shift = F::from(2).pow_vartime([self.limb_width.get() as u64]); // base = 2^limb_width
+        trace!("Shift {shift:?}");
+
+        let mut prev_partial_sum = Option::<AssignedCell<F, F>>::None;
+
+        let limbs = BigUint::from_f(
+            &input.value().unwrap().copied().unwrap_or_default(),
+            self.limb_width,
+            self.limbs_count_limit,
+        )?
+        .limbs()
+        .iter()
+        .map(Some)
+        .chain(iter::repeat(None))
```

Comment:
when we chain the rest with None limb and take limbs_count_limit of them, how do we deal with the None limb here? this one `Value::known(limb.copied().unwrap_or(F::ZERO))`?

IIUC, we always assign limbs_count_limit of cell here, if the number of limbs is less than the limit, we fill zero to the rest cell. Can we just assign number_of_limbs of cells here instead?

Reply:
It is important for us to separate cells with value and cells without values. Look at line 1030 and the expression there. 

In case we always multiply and shift, even when there is no limb - we will shift extra times and the resulting number will not be equal to the target number.

Reply:
After pausing and looking at this code, I thought I'd provalidate whether copy-constraint might be different depending on witness. 

For here it is so, if the number is changed by limb, then fixed will be the same, but copy-constraint on partial sum will be different.

Reply:
if the copy constraint depends on input, we might have some issue. will the unit test catch the issue?

Reply:
I tested this with my hands and halo2 didn't seem to swear at this, but I'm going to try to avoid this situation now.

In general, I can formally copy these columns every time, and the shift is an advice column as it is.

Reply:
Fixed